### PR TITLE
Workflow model

### PR DESCRIPTION
**Pull request by Eleonore9**
Start the household model by creating a `defmodel` and workspace using the api and executor.
- [x] workflow from the model [diagram](https://docs.google.com/drawings/d/1IMUIbpdBhexkfQzm0VUuxIU617TmG9SFPbgx5PMvDNA/edit)
- [x] catalog
- [x] `defmodel` 
- [x] model library
- [x] handle intermediary output (total of households)
- [x] a "skeleton" of empty `definputs` and `defworkflowfns` that match the workflow and catalog + `defworkflowoutput`
- [x] empty placeholder schemas to be replaced later
- [x] add a test to run this "empty" workspace

**Questions:**
-> is the workflow correct (compared to the [diagram](https://docs.google.com/drawings/d/1IMUIbpdBhexkfQzm0VUuxIU617TmG9SFPbgx5PMvDNA/edit))?
-> is there a better way to handle getting the intermediary result (total of household)?
-> is there something else to test for?

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -3,4 +3,8 @@
   :url "http://example.com/FIXME"
   :license {:name "Eclipse Public License"
             :url "http://www.eclipse.org/legal/epl-v10.html"}
-  :dependencies [[org.clojure/clojure "1.7.0"]])
+  :dependencies [[org.clojure/clojure "1.8.0"]
+                 [witan.workspace-api "0.1.20"]
+                 [witan.workspace-executor "0.2.6"
```
Comment (acron0) [inline project.clj]: Put the executor in a test/dev profile so it's only required when running tests.

**Review by acron0 — CHANGES_REQUESTED** [review]
https://github.com/MastodonC/witan.models.demography/blob/master/test/witan/models/acceptance/workspace_test.clj

I think you should change the test to use the style above, rather than the 'run-models' style (`tasks` map and `make-*` functions) as it's way more accurate and less maintainance.

File: test/witan/models/household_test.clj [hunk 2]
```diff
@@ -1,7 +1,57 @@
...
+      (for [{:keys [witan/name witan/fn witan/version witan/params]} catalog
+            :let [fnc (some #(when (and (= fn (:witan/name %))
+                                        (= version (:witan/version %))) %) funs)]]
+        (is fnc)))
+    (testing "The catalog entries are existing functions"
+      (let [library-fns (map #(:witan/impl %) funs)
+            model-ns-list (map str (keys (ns-publics 'witan.models.household)))
+            model-ns-fns (map #(keyword (format "witan.models.household/%s" %)) model-ns-list)]
+        (is (= (set library-fns)
+               (set model-ns-fns)))))))
```
Comment (sunnytownsend) [inline test/witan/models/household_test.clj]: This is the test I remember @acron0 wrote to validate the model - could you run this test also? https://github.com/MastodonC/witan.models.demography/blob/master/test/witan/models/dem/ccm/models_test.clj#L15

**Review by sunnytownsend — COMMENTED** [review]
Remove 'Try and' from beginning of comment - you totally are starting the model!

**Review by sunnytownsend — COMMENTED** [review]
Your question: -> is the workflow correct (compared to the diagram)? 

You could run the graph viz on the workflow and compare against your flowchart

File: test/witan/models/household_test.clj (hunk 2 above)
Comment (Eleonore9) [reply test/witan/models/household_test.clj]: Ok :ok_hand: there are two extra tests here. I'll add them, now!

File: src/witan/models/household.clj [hunk 3]
```diff
@@ -1,6 +1,123 @@
...
+  {:witan/name :hh-model/calc-dwellings
+   :witan/version "1.0.0"
+   :witan/input-schema {:total-households s/TotalHouseholds
+                        :occupancy-rate s/OccupancyRate}
+   :witan/output-schema {:dwellings s/Dwellings}}
+  [{:keys [total-households occupancy-rate]} _]
+  {:dwellings {}})
+
+;; Functions to handle outputs
+(defworkflowfn gather-outputs-1-0-0
```
Comment (sunnytownsend) [inline src/witan/models/household.clj]: Gather has a specific data munging meaning, so please could you change the word `gather` to `collate` or something like that.

File: src/witan/models/model.clj [hunk 4]
```diff
@@ -0,0 +1,111 @@
+(ns witan.models.model
+  (:require [witan.workspace-api :refer [defmodel]]
+            [witan.workspace-api.protocols :as p]
+            [witan.workspace-api.utils :refer [map-fn-meta
+                                               map-model-meta]]
+            [witan.models.household :as hh]))
+
+(def hh-model-workflow
+  [;; Household population
+   [:input-resident-popn :calculate-household-popn]
```
Comment (sunnytownsend) [inline src/witan/models/model.clj:10]: I would suggest editing the words in the flowchart to match the code - so the docs are as clear as poss

File: src/witan/models/model.clj [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+   [:input-institutional-popn :calculate-household-popn]
+   [:calculate-household-popn :group-household-popn]
+
+   ;; Households
+   [:input-household-rates :calculate-households]
```
Comment (sunnytownsend) [inline src/witan/models/model.clj]: I know its a bit anal and makes the fn longer, but there may be different kinds of household rates so could you change `input-household-rates` to `input-household-representative-rates`

File: src/witan/models/model.clj [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+   [:group-household-popn :calculate-households]
+   [:calculate-households :calculate-total-households]
+
+   ;; Household occupancy
+   [:input-vacancy-rates :calculate-occupancy-rate]
+   [:input-second-homes-rates :calculate-occupancy-rate]
+   [:calculate-total-households :calculate-dwellings]
+   [:calculate-occupancy-rate :calculate-dwellings]
+   [:calculate-total-households :gather-outputs]
```
Comment (sunnytownsend) [inline src/witan/models/model.clj]: I guess you should probably add this to your flowchart at least temporarily

File: test/witan/models/household_test.clj (hunk 2 above)
Comment (Eleonore9) [reply test/witan/models/household_test.clj]: That's done now!

File: src/witan/models/household.clj (hunk 3 above)
Comment (Eleonore9) [reply src/witan/models/household.clj]: Done!

File: src/witan/models/model.clj (hunk 5 above)
Comment (Eleonore9) [reply src/witan/models/model.clj]: Done!

**Comment by Eleonore9** [comment]
New diagram: https://docs.google.com/drawings/d/1q_B3zG2EUIIlyREkIjplq5vmiY5U3KUnp-lHtqkExpE/edit

I created a flowchart out of the model workflow and added it to the Readme.

File: project.clj (hunk 1 above)
Comment (Eleonore9) [reply project.clj]: Done

File: src/witan/models/model.clj (hunk 6 above)
Comment (Eleonore9) [reply src/witan/models/model.clj]: I added all the steps to a copy of the flowchart

File: src/witan/models/model.clj (hunk 4 above)
Comment (Eleonore9) [reply src/witan/models/model.clj:10]: Done

**Comment by sunnytownsend** [comment]
excellent stuff! :shipit:
